hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/io/TestHoodieTimelineArchiveLog.java

```diff
@@ -491,6 +519,166 @@ public void testConvertCommitMetadata() {
+
+    createCommitAndRollbackFile("5", "13", false);
+    HoodieInstant notArchivedInstant2 = new HoodieInstant(State.COMPLETED, "rollback", "13");
+
+    HoodieTable table = HoodieSparkTable.create(cfg, context, metaClient);
+    HoodieTimelineArchiveLog archiveLog = new HoodieTimelineArchiveLog(cfg, table);
+
+    archiveLog.archiveIfRequired(context);
+
+    List<HoodieInstant> notArchivedInstants = metaClient.getActiveTimeline().reload().getRollbackTimeline().getInstants().collect(Collectors.toList());
+    //There will be 2 * 2 files but due to TimelineLayoutV1 this will show as 2.
+    assertEquals(2, notArchivedInstants.size(), "Not archived instants should be 2");
+    assertEquals(notArchivedInstants, Arrays.asList(notArchivedInstant1, notArchivedInstant2), "");
+  }
+
+  @Test
+  public void testArchiveCompletedShouldRetainMinInstantsIfInstantsGreaterThanMaxtoKeep() throws IOException {
+    int minInstants = 2;
+    int maxInstants = 10;
+    HoodieWriteConfig cfg =
+            HoodieWriteConfig.newBuilder().withPath(basePath).withSchema(HoodieTestDataGenerator.TRIP_EXAMPLE_SCHEMA)
+                    .withParallelism(2, 2).forTable("test-trip-table")
+                    .withCompactionConfig(HoodieCompactionConfig.newBuilder().retainCommits(1).archiveCommitsWith(minInstants, maxInstants).build())
+                    .build();
+    metaClient = HoodieTableMetaClient.reload(metaClient);
+    for (int i = 0; i < maxInstants + 2; i++) {
+      createCleanMetadata(i + "", false);
+    }
+
+    HoodieTable table = HoodieSparkTable.create(cfg, context, metaClient);
+    HoodieTimelineArchiveLog archiveLog = new HoodieTimelineArchiveLog(cfg, table);
+
+    archiveLog.archiveIfRequired(context);
+    assertEquals(minInstants, metaClient.getActiveTimeline().reload().getInstants().count());
+  }
+
+  @Test
+  public void testArchiveCompletedShouldNotArchiveIfInstantsLessThanMaxtoKeep() throws IOException {
+    int minInstants = 2;
+    int maxInstants = 10;
+    HoodieWriteConfig cfg =
+            HoodieWriteConfig.newBuilder().withPath(basePath).withSchema(HoodieTestDataGenerator.TRIP_EXAMPLE_SCHEMA)
+                    .withParallelism(2, 2).forTable("test-trip-table")
+                    .withCompactionConfig(HoodieCompactionConfig.newBuilder().retainCommits(1).archiveCommitsWith(minInstants, maxInstants).build())
+                    .build();
+    metaClient = HoodieTableMetaClient.reload(metaClient);
+    for (int i = 0; i < maxInstants; i++) {
+      createCleanMetadata(i + "", false);
+    }
+
+    HoodieTable table = HoodieSparkTable.create(cfg, context, metaClient);
+    HoodieTimelineArchiveLog archiveLog = new HoodieTimelineArchiveLog(cfg, table);
+
+    archiveLog.archiveIfRequired(context);
+    assertEquals(maxInstants, metaClient.getActiveTimeline().reload().getInstants().count());
+  }
+
+  @Test
+  public void testArchiveCompletedRollbackAndClean() throws IOException {
```

Comment:
I see a lot of tests which is good. But if i read correctly, some of these are repetitive. Can you comment this test and see if coverage changes? CI has been a pain, so we want to make sure to keep tests that only add value.

Reply:
This one is specifically to add/show that max and min InstantsToKeep limits are applied at an action level for completed instants.